packages/@ember/-internals/glimmer/lib/helpers/action.ts

```diff
@@ -298,16 +298,23 @@ export default internalHelper((args: CapturedArguments): Reference<Function> =>
   } else {
     fn = makeDynamicClosureAction(
       valueForRef(context) as object,
-      target!,
-      action,
+      // SAFETY: glimmer-vm should expose narrowing utilities for references
+      //         as is, `target` is still `Reference<unknown>`.
+      //         however, we never even tried to narrow `target`, so this is potentially risky code.
+      target! as Reference<MaybeActionHandler>,
+      // SAFETY: glimmer-vm should expose narrowing utilities for references
+      //         as is, `action` is still `Reference<unknown>`
+      action as Reference<string | AnyFn>,
```

Comment:
I think this is just exposing a real bug, we really have no business making these unfounded claims here. We should keep them as `Reference<unknown>` and push it onto the consumption side to actually narrow the type with debug assertions when they have actual values.

Reply:
well, isn't `isInvokableRef(action)` supposed to narrow to `string | AnyFn`?

Reply:
Getting rid of the casts here causes:
![image](https://github.com/emberjs/ember.js/assets/199018/051facf1-7486-4297-85c5-48f4a32dca50)

which... feels worse?

In any case, if this is a real bug, I think it should be handled in a separate PR as it's been around for a long while, and only exposed through glimmer-vm's new stricter types.

Reply:
In general, we don't know what is _in_ a reference without dereferencing it, just like there is no correct way to implement `isFunctionOfString(func: () => unknown): func is () => string` without actually calling it, unless we are sure that we were the one that made the reference in the first place and specifically know what is in it, which doesn't seem to be the case here.

What I am saying is that, since we don't know what is in it, we should keep it as `Reference<unknown>` on this line, and 100% the intention is specifically to _cause_ that TypeError you found. This identifies the place where we are currently making an incorrect claim.

You then change those to `Reference<unknown>` as well, which pushes the type error further down, to the line where we currently, incorrectly, does `let something: TButProbablyInferred = valueForRef(ref);`. Now here is the real issue, but this is where the place we can actually do something about it, since, we now have the actual dereferenced value:

```ts
let something = valueForRef(ref); // now something is correctly `unknown`
assert("something must be a string!", typeof something === "string");
// now something is `string`
```

Reply:
Fixed in 510d28c01